{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "77c3eaea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter\n",
    "import sys\n",
    "import os\n",
    "\n",
    "if os.environ.get('DISPLAY','') == '':\n",
    "    print('no display found. Using :0.0')\n",
    "    os.environ.__setitem__('DISPLAY', ':0.0')\n",
    "\n",
    "if not os.getcwd().endswith('Scripts'):\n",
    "    os.chdir('Scripts')\n",
    "    \n",
    "import PySimpleGUI as sg\n",
    "import functions as fxn\n",
    "import stitchr as st\n",
    "import collections as coll\n",
    "\n",
    "\n",
    "# #create main window\n",
    "# master = tkinter.Tk()\n",
    "# master.title(\"tester\")\n",
    "# master.geometry(\"300x100\")\n",
    "\n",
    "\n",
    "# #make a label for the window\n",
    "# label1 = tkinter.Label(master, text='Hellooooo')\n",
    "# # Lay out label\n",
    "# label1.pack()\n",
    "\n",
    "# # Run forever!\n",
    "# master.mainloop()\n",
    "\n",
    "extra_gene_text = \"(FASTA format, optional)\"\n",
    "\n",
    "col1=[\n",
    "    \n",
    "    \n",
    "    [sg.Button('Example data'), sg.Button('Reset form')],\n",
    "    \n",
    "    [sg.Text('Species', size=(30, 1), font=(\"Helvetica\", 12))],    \n",
    "    [sg.Radio('Human', \"RADIO1\", key='rad_hs', default=True), sg.Radio('Mouse', \"RADIO1\", key='rad_mm')],\n",
    "    \n",
    "    [sg.Text('Additional genes', size=(30, 1), font=(\"Helvetica\", 12))],\n",
    "    [sg.MLine(default_text=extra_gene_text, size=(35, 3), key='additional_genes')],\n",
    "    \n",
    "    [sg.Button('Run Stitchr')],\n",
    "    \n",
    "    [sg.Button('Exit')]\n",
    "    \n",
    "    \n",
    "]\n",
    "\n",
    "\n",
    "col2=[\n",
    "    \n",
    "    [sg.Text('TRA')],\n",
    "\n",
    "    [sg.Text('TRAV')], [sg.InputText('', key='TRAV')],\n",
    "\n",
    "    [sg.Text('TRAJ')],[sg.InputText('', key='TRAJ')],\n",
    "\n",
    "    [sg.Text('TRA CDR3')], [sg.InputText('', key='TRA_cdr3')],\n",
    "    \n",
    "    [sg.Text('TRA out')], [sg.MLine(default_text='', size=(50, 20),\n",
    "              key='TRA_out')]\n",
    "      \n",
    "]\n",
    "\n",
    "\n",
    "col3=[\n",
    "    \n",
    "    [sg.Text('TRB')],\n",
    "      \n",
    "    [sg.Text('TRBV')], [sg.InputText('', key='TRBV')],\n",
    "\n",
    "    [sg.Text('TRBJ')], [sg.InputText('', key='TRBJ')],\n",
    "\n",
    "    [sg.Text('TRB CDR3')], [sg.InputText('', key='TRB_cdr3')],\n",
    "    \n",
    "    [sg.Text('TRB out')], [sg.MLine(default_text='', size=(50, 20),\n",
    "              key='TRB_out')]\n",
    "      \n",
    "]\n",
    "\n",
    "\n",
    "layout = [[sg.Column(col1, element_justification='l'), \n",
    "           sg.Column(col2, element_justification='l'),\n",
    "           sg.Column(col3, element_justification='l')]]\n",
    "\n",
    "window = sg.Window(\"stitchr\",layout)\n",
    "#event, values = window.read()\n",
    "\n",
    "example_data = {\n",
    "    'TRAV': 'TRAV1-2',\n",
    "    'TRAJ': 'TRAJ33',\n",
    "    'TRA_cdr3': 'CAVLDSNYQLIW',\n",
    "    'TRA_out': '',\n",
    "    'TRBV': 'TRBV7-3*01',\n",
    "    'TRBJ': 'TRBJ1-1*01',\n",
    "    'TRB_cdr3': 'CASSYLQAQYTEAFF',\n",
    "    'TRB_out': '',\n",
    "}\n",
    "\n",
    "while True:\n",
    "    event, values = window.read()\n",
    "\n",
    "    if event == 'Example data':\n",
    "    \n",
    "        for field in example_data:\n",
    "            window[field].update(example_data[field])\n",
    "    \n",
    "    elif event == 'Reset form':\n",
    "    \n",
    "        for field in example_data:\n",
    "            window[field].update('')\n",
    "                \n",
    "    elif event == 'Run Stitchr':\n",
    "        \n",
    "        # Disable stitchr button while code is running\n",
    "        window['Run Stitchr'].update(disabled=True)\n",
    "    \n",
    "        # Determine species \n",
    "        if values['rad_hs']:\n",
    "            species = 'HUMAN'\n",
    "        elif values['rad_mm']:\n",
    "            species = 'MOUSE'\n",
    "        else:\n",
    "            raise IOError('Species cannot be determined')\n",
    "        \n",
    "        print(species)\n",
    "        \n",
    "        # Loop through both chains, determine which are asked for, and read data in\n",
    "        codons = fxn.get_optimal_codons('../Data/' + species + '/kazusa.txt', species)\n",
    "#        tcr_dat, functionality, out_lists, stitched, offsets, out_strs = [{}] * 6 # Todo rm\n",
    "        outputs = coll.defaultdict()\n",
    "    \n",
    "        for chain in ['TRA', 'TRB']:\n",
    "            \n",
    "            if values[chain + 'V'] and values[chain + 'J'] and values[chain + '_cdr3']:\n",
    "                \n",
    "                tcr_dat, functionality = fxn.get_imgt_data(chain, st.gene_types, species)\n",
    "                \n",
    "                print(chain)\n",
    "               # print(tcr_dat)\n",
    "                \n",
    "                \n",
    "                tcr_bits = {'v': values[chain + 'V'], 'j': values[chain + 'J'], 'cdr3': values[chain + '_cdr3'],\n",
    "                            'skip_c_checks': False, 'species': species, 'name': ''}\n",
    "                tcr_bits = fxn.autofill_input(tcr_bits, chain)\n",
    "\n",
    "                print(tcr_bits)\n",
    "                \n",
    "                # Run the stitching   \n",
    "#                 out_lists[chain], stitched[chain], offsets[chain] = st.stitch(\n",
    "                outputs[chain + '_out_list'], outputs[chain + '_stitched'], outputs[chain + '_offset'] = st.stitch(\n",
    "                    tcr_bits, chain, tcr_dat, functionality, codons, 3)\n",
    "                \n",
    "                outputs[chain + '_out_str'] =  '|'.join(outputs[chain + '_out_list'])\n",
    "                outputs[chain + '_fasta'] = fxn.fastafy('nt|' + outputs[chain + '_out_str'], \n",
    "                                                        outputs[chain + '_stitched'])\n",
    "                \n",
    "                window[chain + '_out'].update(outputs[chain + '_fasta'])\n",
    "                \n",
    "#                 # Use the offset to 5' pad the stitched sequence with 'N's to make up for non-codon length 5' added sequences\n",
    "#                 print(fxn.fastafy('aa|' + out_str[chain], fxn.translate_nt('N' * offset[chain] + stitched[chain])))        \n",
    "        \n",
    "        print('Stitching completed')\n",
    "        \n",
    "        # Re-enable stitchr button once completed\n",
    "        window['Run Stitchr'].update(disabled=False)\n",
    "\n",
    "        \n",
    "    elif event in ('Exit', None):\n",
    "        break\n",
    "\n",
    "window.close()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7035701",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
